contextualized/dags/torch_notmad/torch_notmad.py

```diff
@@ -0,0 +1,286 @@
+                 archetype_loss_params = {'l1': 0., 'alpha': 1e-1, 'rho': 1e-2},
+                 opt_lr = 1e-3,
+                 opt_step= 50,
+                 n_encoder_hidden_layers=2,
+                 encoder_width=32,
+                 init_mat=None,
+                 auto_opt=True,
+                 encoder_type="NGAM"
+            ):
+        """ Initialize NOTMAD.
+        
+        Args:
+            datamodule (pl.LightningDataModule): Lightning datamodule to use for training
+        
+        Kwargs:
+            Explainer Kwargs
+            ----------------
+            init_mat (np.array): 3D Custom initial weights for each archetype. Defaults to None.
+            n_archetypes (int:4): Number of archetypes in explainer
+            
+            Encoder Kwargs
+            ----------------
+            n_encoder_hidden_layers(int:2): Number encoder hidden layers
+            encoder_width(int:32): Width of encoder hidden layers
+            encoder_type (str: NGAM): Encoder module to use
+            
+            Optimization Kwargs
+            -------------------
+            auto_opt(bool: True): Use torch's backprop vs manual (customizable backprop in self.training_step)
+            opt_lr(float): Optimizer learning rate
+            opt_step(int): Optimizer step size
+            
+            Loss Kwargs
+            -----------
+            sample_specific_loss_params (dict of str: int): Dict of params used by NOTEARS loss (l1, alpha, rho)
+            archetype_specific_loss_params (dict of str: int): Dict of params used by Archetype loss (l1, alpha, rho)
+
+        """
+        super(NOTMAD_model, self).__init__()
+
+        #dataset params
+        self.datamodule = datamodule
+        self.n_archetypes = n_archetypes
+        self.context_shape = self.datamodule.C_train.shape
+        self.feature_shape = self.datamodule.X_train.shape
+
+        #dag/loss params
+        self.project_distance = 0.1
+        self.archetype_loss_params = archetype_loss_params
+        self.alpha, self.rho, self.use_dynamic_alpha_rho = self._parse_alpha_rho(sample_specific_loss_params)
+
+        #torch params
+        self.opt_lr = opt_lr
+        self.opt_step = opt_step
+        self.automatic_optimization = auto_opt
+        self.encoder_type = encoder_type
+
+        #layer shapes 
+        encoder_input_shape = (self.context_shape[1], 1)
+        encoder_output_shape = (self.n_archetypes, )
+        arch_shape = (self.n_archetypes, self.feature_shape[-1])
+        
+        #layer params
+        self.init_mat = init_mat
+
+        #layers
+        if self.encoder_type == "NGAM":
+            self.encoder = NGAM(encoder_input_shape[0], encoder_output_shape[0],
+                                n_hidden_layers=n_encoder_hidden_layers, 
+                                width=encoder_width)
+        else:
+            self.encoder = self._build_encoder(encoder_input_shape[0], 
```

Comment:
_build_encoder is depreciated, should specify encoders from modules.py and allow users to pass encoder_kwargs. See regression.metamodels.py for examples of how this might work.